src/hotspot/share/gc/shenandoah/shenandoahDegeneratedGC.cpp

```diff
@@ -280,38 +280,17 @@ void ShenandoahDegenGC::op_degenerated() {
       // In above case, update roots should disarm them
       ShenandoahCodeRoots::disarm_nmethods();
 
-      if (heap->mode()->is_generational() && heap->is_concurrent_old_mark_in_progress()) {
-        // This is still necessary for degenerated cycles because the degeneration point may occur
-        // after final mark of the young generation. See ShenandoahConcurrentGC::op_final_updaterefs for
-        // a more detailed explanation.
-        heap->old_generation()->transfer_pointers_from_satb();
-      }
-
       op_cleanup_complete();
-      // We defer generation resizing actions until after cset regions have been recycled.
+
       if (heap->mode()->is_generational()) {
-        auto result = ShenandoahGenerationalHeap::heap()->balance_generations();
-        LogTarget(Info, gc, ergo) lt;
-        if (lt.is_enabled()) {
-          LogStream ls(lt);
-          result.print_on("Degenerated GC", &ls);
-        }
+        ShenandoahGenerationalHeap::heap()->complete_degenerated_cycle();
```

Comment:
Not a change that you made, but I wondered about the semantics of "complete" in the name "op_cleanup_complete". It appears to use a phase timer around its work. Was the idea to include the "complete_degenerated_cycle" work inside that phase for timers associated with whatever `ShenandoahPhaseTimings::degen_gc_cleanup_complete` wanted to track, or is it good as is to exclude that generational work which may be is separately tracked?

Reply:
(fwiw, I didn't see any obvious phase timers around the work in the generational `complete_degenerated_cycle`. Should there be any?)

Reply:
We have a phase timer for coalesce and fill:
```C
    ShenandoahGCPhase phase(ShenandoahPhaseTimings::degen_gc_coalesce_and_fill);
```
We could introduce a timer around this extracted method, but I expect coalesce and fill will be the long pole.